File: frontend/src/lib/payrollRemarks.ts
export type PayrollRemarksLayout = {
  max_lines: number;
  max_width: number;
  max_length: number;
  character_widths: Record<string, number>;
  unknown_width: number;
};

export type PayrollRemarks = {
  remarks: string;
  editable: boolean;
  layout: PayrollRemarksLayout;
};

export function normalizePayrollRemarks(value: string): string {
  return value.replace(/\r\n?/g, "\n").replace(/\t/g, "    ");
}

// Same wrapping as the exporter. The API supplies the printed font metrics.
export function wrapPayrollRemarks(value: string, layout: PayrollRemarksLayout): string[] {
  value = normalizePayrollRemarks(value);
  if ([...value].some((char) => {
    const code = char.codePointAt(0)!;
    return (code < 32 && char !== "\n") || (code >= 0xd800 && code <= 0xdfff)
      || code === 0xfffe || code === 0xffff;
  })) {
    throw new Error("Die Bemerkung enthält ein nicht unterstütztes Steuerzeichen.");
  }
  if (!value) return [];
  const lines: string[] = [];
  for (const raw of value.split("\n")) {
    let paragraph = [...raw];
    while (paragraph.length) {
      let width = 0;
      let end = 0;
      for (const char of paragraph) {
        const nextWidth = width + (layout.character_widths[char] ?? layout.unknown_width);
        if (nextWidth > layout.max_width) break;
        width = nextWidth;
        end += 1;
      }
      if (end === paragraph.length) break;
      const space = paragraph.slice(0, end).lastIndexOf(" ");
      const cut = space > 0 ? space + 1 : end;
      lines.push(paragraph.slice(0, cut).join(""));
      paragraph = paragraph.slice(cut);
    }
    lines.push(paragraph.join(""));
  }
  return lines;
}

export function payrollRemarksError(value: string, layout: PayrollRemarksLayout): string | null {
  try {
    if ([...normalizePayrollRemarks(value)].length > layout.max_length
      || wrapPayrollRemarks(value, layout).length > layout.max_lines) {
      return `Das Excel-Feld hat nur ${layout.max_lines} Zeilen. Bitte die Bemerkung kürzen.`;
    }
    return null;
  } catch (error) {
    return error instanceof Error ? error.message : "Die Bemerkung ist ungültig.";
  }
}

File: backend/app/services/payroll_remarks.py
"""Four printed lines in I46:L49, measured for Arial 10 with a safety margin."""
from reportlab.pdfbase.pdfmetrics import stringWidth

MAX_LINES = 4
MAX_WIDTH = 14400  # 144 pt at 10 pt; merged template cells are about 162 pt wide.
MAX_LENGTH = 512
CHARACTER_WIDTHS = {
    char: round(stringWidth(char, "Helvetica", 1000))
    for char in dict.fromkeys("".join(chr(n) for n in range(32, 256)) + "€„“‚‘’–—…")
}
UNKNOWN_WIDTH = 2000  # Conservative allowance for fallback fonts / Unicode symbols.


def remarks_layout() -> dict:
    return {"max_lines": MAX_LINES, "max_width": MAX_WIDTH,
            "max_length": MAX_LENGTH, "character_widths": CHARACTER_WIDTHS,
            "unknown_width": UNKNOWN_WIDTH}


def normalize_remarks(value: str) -> str:
    return value.replace("\r\n", "\n").replace("\r", "\n").replace("\t", "    ")


def wrap_remarks(value: str) -> list[str]:
    """Keep explicit breaks; word-wrap and split long words without losing text."""
    value = normalize_remarks(value)
    if any(ord(c) < 32 and c != "\n" or 0xD800 <= ord(c) <= 0xDFFF
           or ord(c) in (0xFFFE, 0xFFFF) for c in value):
        raise ValueError("Die Bemerkung enthält ein nicht unterstütztes Steuerzeichen.")
    if not value:
        return []
    lines = []
    for paragraph in value.split("\n"):
        while paragraph:
            width, end = 0, 0
            for char in paragraph:
                next_width = width + CHARACTER_WIDTHS.get(char, UNKNOWN_WIDTH)
                if next_width > MAX_WIDTH:
                    break
                width, end = next_width, end + 1
            if end == len(paragraph):
                break
            space = paragraph.rfind(" ", 0, end)
            cut = space + 1 if space > 0 else end
            lines.append(paragraph[:cut])
            paragraph = paragraph[cut:]
        lines.append(paragraph)
    return lines


def validate_remarks(value: str) -> str:
    value = normalize_remarks(value)
    if len(value) > MAX_LENGTH or len(wrap_remarks(value)) > MAX_LINES:
        raise ValueError("Das Excel-Feld hat nur 4 Zeilen. Bitte die Bemerkung kürzen.")
    return value
